Docs: YesSql (#19594)
OrchardCMS/OrchardCore: src/docs/topics/data/yessql.md

# How YesSql works

> Orchard Core does not use an ORM like Entity Framework. All its data — content items, users, settings, workflows... — is stored with [YesSql](https://github.com/sebastienros/yessql), a .NET **document database interface over a relational database**. You get the flexibility of a document store (no schema to migrate when your objects change) while still running on SQL Server, SQLite, MySQL or PostgreSQL.

Reviewer: ```suggestion
Orchard Core does not use an ORM like Entity Framework. All its data — content items, users, settings, workflows... — is stored with [YesSql](https://github.com/sebastienros/yessql), a .NET **document database interface over a relational database**. You get the flexibility of a document store (no schema to migrate when your objects change) while still running on SQL Server, SQLite, MySQL, or PostgreSQL.
```

## Documents

When an object is saved, YesSql serializes it to JSON and stores it as a row in a single `Document` table (per tenant, with an optional table prefix):

| Id | Type | Content | Version |
| --- | --- | --- | --- |
| 42 | `OrchardCore.ContentManagement.ContentItem, OrchardCore.ContentManagement.Abstractions` | `{ "ContentItemId": "4tavbc...", "DisplayText": "My blog post", ... }` | 3 |

This is why adding a field or a part to a content type in Orchard Core never requires a database migration: the shape of the JSON just changes.

The obvious limitation is that you cannot efficiently query inside a JSON blob with SQL. That is what indexes are for.

## Indexes

An index is a plain C# class holding the properties you want to query on. YesSql maintains one **regular SQL table per index type**, with a `DocumentId` column pointing back to the `Document` table. When a document is created, updated, or deleted, its index rows are recomputed within the same transaction.

There are two kinds:

> - **`MapIndex`**: one (or several) index rows per document. For example `ContentItemIndex` maps every content item to its `ContentType`, `Published`, `Owner`, etc., and `AliasPartIndex` maps every content item that has an `AliasPart` to its alias.

Reviewer: ```suggestion
- **`MapIndex`**: one (or several) index rows per document. For example, `ContentItemIndex` maps every content item to its `ContentType`, `Published`, `Owner`, etc., and `AliasPartIndex` maps every content item that has an `AliasPart` to its alias.
```

> - **`MapIndex`**: one (or several) index rows per document. For example `ContentItemIndex` maps every content item to its `ContentType`, `Published`, `Owner`, etc., and `AliasPartIndex` maps every content item that has an `AliasPart` to its alias.

Reviewer: `MapIndex` maps to one object only

Reviewer 2: Clarified this to distinguish row ownership from mapping cardinality: each MapIndex row maps to one document, while a mapping can emit multiple rows for a document.

> - **`ReduceIndex`**: one row per group of documents, aggregating values (like a `GROUP BY`). Used for counting or grouping scenarios.

Reviewer: `ReduceIndex` is a one-to-many mapping; that's why it called reduce

Reviewer 2: Clarified this as many documents reducing into one grouped index row, consistent with YesSql ReduceIndex semantics.

## The session

All reads and writes go through `ISession`, a unit of work registered in the dependency injection container (one per request). `Save` and `Delete` are buffered in memory; the SQL commands run in a single transaction that is committed at the end of the request, or when `SaveChangesAsync` is called explicitly.

```csharp
public sealed class MyController : Controller
{
    private readonly ISession _session;

    public MyController(ISession session)
    {
        _session = session;
    }

    public async Task<IActionResult> Cheap()
    {
        // Query documents through an index.
        var cheapProducts = await _session
            .Query<ContentItem, ProductIndex>(index => index.Price < 10)
            .ListAsync();

        return View(cheapProducts);
    }
}
```

> The query filters on the index table with regular SQL, then loads the matching JSON documents and deserializes them. Documents are cached by the session: loading the same document twice in a request returns the same instance.

Reviewer: ```suggestion
The query filters on the index table with regular SQL, then loads the matching JSON documents and deserializes them. The session caches documents: loading the same document twice in a request returns the same instance.
```

> For content items specifically, prefer the higher-level `IContentManager` (or `IOrchardHelper` extensions like `QueryContentItemsAsync`) which take care of loading, versioning, and handlers; drop down to `ISession` when you need to query on your own indexes.

Reviewer: ```suggestion
For content items specifically, prefer the higher-level `IContentManager` (or `IOrchardHelper` extensions like `QueryContentItemsAsync`), which handle loading, versioning, and handlers; drop down to `ISession` when you need to query your own indexes.
```
